hpx/parallel/algorithms/sort.hpp

```diff
@@ -0,0 +1,372 @@
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Compare, typename Proj>
+        struct compare_projected
+        {
+            template <typename Compare_, typename Proj_>
+            compare_projected(Compare_ && comp, Proj_ && proj)
+              : comp_(std::forward<Compare>(comp)),
+                proj_(std::forward<Proj>(proj))
+            {}
+
+            template <typename T>
+            inline bool operator()(T const& t1, T const& t2)
+            {
+                return hpx::util::invoke(comp_,
+                    hpx::util::invoke(proj_, t1),
+                    hpx::util::invoke(proj_, t2));
+            }
+
+            Compare comp_;
+            Proj proj_;
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        // std::is_sorted is not available on all supported platforms yet
+        template <typename Iter, typename Compare>
+        inline bool is_sorted_sequential(Iter first, Iter last, Compare comp)
+        {
+            bool sorted = true;
+            if (first != last)
+            {
+                for (Iter it1 = first, it2 = first + 1;
+                     it2 != last && (sorted = !comp(*it2, *it1));
+                     it1 = it2++)
+                {
+                    /**/
+                }
+            }
+            return sorted;
+        }
+
+        //------------------------------------------------------------------------
+        //  function : sort_thread
+        /// \brief this function is the work assigned to each thread in the
+        ///        parallel process
+        /// \exception
+        /// \return
+        /// \remarks
+        //------------------------------------------------------------------------
+        template <typename ExPolicy, typename RandomIt, typename Compare>
+        hpx::future<void> sort_thread(ExPolicy policy,
+            RandomIt first, RandomIt last, Compare comp)
+        {
+            typedef typename ExPolicy::executor_type executor_type;
+            typedef typename hpx::parallel::executor_traits<executor_type>
+                executor_traits;
+
+            //------------------------- begin ----------------------
+            std::ptrdiff_t N = last - first;
+            if (std::size_t(N) <= sort_limit_per_task)
+            {
+                return executor_traits::async_execute(
+                    policy.executor(),
+                    [first, last, comp]()
+                    {
+                        std::sort(first, last, comp);
+                    });
+            }
+
+            //----------------------------------------------------------------
+            //                     split
+            //----------------------------------------------------------------
+
+            //------------------- check if sorted ----------------------------
+            if (detail::is_sorted_sequential(first, last, comp))
```

Comment:
After more thinking - I believe the `is_sorted` test should be done for the leaves only (before line 101 above?). Currently the algorithm re-scans the input data over and over again. This can be a big hit for nearly sorted data, mainly.

The current code assumes that testing the data sequence for being already sorted adds less overheads than spawning the two sub-tasks for the two partitions. This might be a valid assumption for kernel threads, it might not be a correct assessment for HPX threads, however.